rhp/v4/server.go

```diff
@@ -0,0 +1,1078 @@
+	settings.Prices.ValidUntil = time.Now().Add(s.priceTableValidity)
+	sigHash := settings.Prices.SigHash()
+	settings.Prices.Signature = s.hostKey.SignHash(sigHash)
+
+	return rhp4.WriteResponse(stream, &rhp4.RPCSettingsResponse{
+		Settings: settings,
+	})
+}
+
+func (s *Server) handleRPCReadSector(stream net.Conn) error {
+	var req rhp4.RPCReadSectorRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	prices, token := req.Prices, req.Token
+	if err := prices.Validate(s.hostKey.PublicKey()); err != nil {
+		return errorBadRequest("price table invalid: %v", err)
+	} else if err := token.Validate(); err != nil {
+		return errorBadRequest("account token invalid: %v", err)
+	} else if err := req.Validate(); err != nil {
+		return errorBadRequest("request invalid: %v", err)
+	}
+
+	if err := s.contractor.DebitAccount(req.Token.Account, prices.RPCReadSectorCost(req.Length)); err != nil {
+		return fmt.Errorf("failed to debit account: %w", err)
+	}
+
+	sector, err := s.sectors.ReadSector(req.Root)
+	if err != nil {
+		return fmt.Errorf("failed to read sector: %w", err)
+	}
+
+	segment := sector[req.Offset : req.Offset+req.Length]
+
+	return rhp4.WriteResponse(stream, &rhp4.RPCReadSectorResponse{
+		Sector: segment,
+		Proof:  nil, // TODO implement proof
+	})
+}
+
+func (s *Server) handleRPCWriteSector(stream net.Conn) error {
+	var req rhp4.RPCWriteSectorStreamingRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+	prices, token := req.Prices, req.Token
+	if err := prices.Validate(s.hostKey.PublicKey()); err != nil {
+		return errorBadRequest("price table invalid: %v", err)
+	} else if err := token.Validate(); err != nil {
+		return errorBadRequest("account token invalid: %v", err)
+	}
+
+	settings := s.settings.RHP4Settings()
+
+	switch {
+	case req.DataLength > rhp4.SectorSize:
+		return errorBadRequest("sector size %v exceeds maximum %v", req.DataLength, rhp4.SectorSize)
+	case req.DataLength%rhp4.LeafSize != 0:
+		return errorBadRequest("sector length %v must be a multiple of segment size %v", req.DataLength, rhp4.LeafSize)
+	case req.Duration > settings.MaxSectorDuration:
+		return errorBadRequest("sector duration %v exceeds maximum %v", req.Duration, settings.MaxSectorDuration)
+	case settings.RemainingStorage < rhp4.SectorSize:
+		return rhp4.ErrNotEnoughStorage
+	}
+
+	// read the sector data
+	var sector [rhp4.SectorSize]byte
+	if _, err := io.ReadFull(stream, sector[:req.DataLength]); err != nil {
+		return errorDecodingError("failed to read sector data: %v", err)
+	}
+
+	// calculate the cost
+	cost := prices.RPCWriteSectorCost(req.DataLength, req.Duration)
+	// debit the account
+	if err := s.contractor.DebitAccount(req.Token.Account, cost); err != nil {
+		return fmt.Errorf("failed to debit account: %w", err)
+	}
+
+	// store the sector
+	root := rhp4.SectorRoot(&sector)
```

Comment:
this should be fused with the `ReadFull` above (via `ReaderRoot`?) for performance